### PR TITLE
PLFM-6676

### PR DESCRIPTION
**Pull request by xschildw**

### CODE REVIEW

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/dataaccess/AccessRequirementManagerImpl.java [hunk 1]
```diff
@@ -232,6 +254,7 @@ public AccessRequirement getAccessRequirement(String requirementId) throws Datas
 		for (RestrictableObjectDescriptor rod : toUpdate.getSubjectIds()) {
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/dataaccess/AccessRequirementManagerImpl.java]: We need a delta of the current subjects and the updated subjects to capture the case where a subject ID is removed (the main use case).

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/dataaccess/AccessRequirementManagerImplUnitTest.java [hunk 2]
```diff
@@ -480,6 +510,7 @@ public void testUpdateConcreteType() {
 		Assertions.assertThrows(IllegalArgumentException.class, () -> {
 			arm.updateAccessRequirement(userInfo, accessRequirementId, toUpdate);
 		});
+		verify(mockTransactionalMessenger).sendMessageAfterCommit(TEST_ENTITY_ID, ObjectType.ENTITY, ChangeType.UPDATE);
```
Comment (john-hill) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/dataaccess/AccessRequirementManagerImplUnitTest.java]: missing case where there are multiple subjects and case where subject is a folder or project.  Will need a test for the delta on update.

**Review by john-hill — CHANGES_REQUESTED** [review]

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/dataaccess/AccessRequirementManagerImplUnitTest.java [hunk 3]
```diff
@@ -956,4 +1084,81 @@ public void testValidateAccessRequirement() {
...
+	@Test
+	public void testSignalDeletedSubjectIdsOneDeleted() {
+		List<RestrictableObjectDescriptor> currentRods = generateRods(1); // id == 0
+		List<RestrictableObjectDescriptor> updatedRods = new ArrayList<>();
+		RestrictableObjectDescriptor rod = new RestrictableObjectDescriptor();
+		rod.setId("1");
+		rod.setType(RestrictableObjectType.ENTITY);
+		updatedRods.add(rod);
+		when(nodeDao.getNodeTypeById("0")).thenThrow(new NotFoundException());
+		// call under test: "0" should all be signaled
```
Comment (xschildw) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/dataaccess/AccessRequirementManagerImplUnitTest.java]: Comment is incorrect. In this case we catch the NotFound and return b4 sending msg.

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/dataaccess/AccessRequirementManagerImpl.java [hunk 4]
```diff
@@ -124,6 +132,33 @@ private void preventCreateWithinSTSFolder(String entityId) {
 		}
 	}
 
+	private void signalSubjectId(RestrictableObjectDescriptor rod) {
```
Comment (xschildw) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/dataaccess/AccessRequirementManagerImpl.java]: Should not be private to allow for testing

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/dataaccess/AccessRequirementManagerImpl.java [hunk 5]
```diff
@@ -1,5 +1,6 @@
 package org.sagebionetworks.repo.manager.dataaccess;
 
+import java.nio.channels.NotYetBoundException;
```
Comment (xschildw) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/dataaccess/AccessRequirementManagerImpl.java:3]: Delete

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/dataaccess/AccessRequirementManagerImplAutoWiredTest.java [hunk 6]
```diff
@@ -139,6 +139,16 @@ public void before() throws Exception {
 
 	@AfterEach
 	public void after() throws Exception {
+
+		if (ar!=null && ar.getId()!=null && accessRequirementManager!=null) {
+			//
+			try {
+				accessRequirementManager.deleteAccessRequirement(adminUserInfo, ar.getId().toString());
```
Comment (john-hill) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/dataaccess/AccessRequirementManagerImplAutoWiredTest.java:146]: Remove is not needed.

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/dataaccess/AccessRequirementManagerImpl.java [hunk 7]
```diff
@@ -132,12 +181,12 @@ private void preventCreateWithinSTSFolder(String entityId) {
 			throw new UnauthorizedException("Only ACT member can create an AccessRequirement.");
 		}
 		populateCreationFields(userInfo, accessRequirement);
-		
 		for (RestrictableObjectDescriptor rod : accessRequirement.getSubjectIds()) {
 			if (RestrictableObjectType.ENTITY==rod.getType()) {
 				preventCreateWithinSTSFolder(rod.getId());
 			}
 		}
+		signalSubjectIds(new ArrayList<RestrictableObjectDescriptor>(), accessRequirement.getSubjectIds()); // Create == empty currentSibjectIds
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/dataaccess/AccessRequirementManagerImpl.java:189]: Collections.emptyList()

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/dataaccess/AccessRequirementManagerImplUnitTest.java [hunk 8]
```diff
@@ -956,4 +1083,150 @@ public void testValidateAccessRequirement() {
...
+
+		verify(nodeDao).getNodeTypeById("0");
+		verify(mockTransactionalMessenger).sendMessageAfterCommit("0", ObjectType.ENTITY, ChangeType.UPDATE);
+	}
+
+	@Test
+	public void testFindSubjectIdsToSignalCurrentEmpty() {
+		Set<RestrictableObjectDescriptor> currentSubjectIds = new HashSet<>();
+		Set<RestrictableObjectDescriptor> updatedSubjectIds = generateSubjectIds(2, 2); // 2,3
+		Set<RestrictableObjectDescriptor> subjectIdsToSignal = arm.findSubjectIdsToSignal(currentSubjectIds, updatedSubjectIds);
```
Comment (john-hill) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/dataaccess/AccessRequirementManagerImplUnitTest.java:1154]: // call under test

**Review by john-hill — APPROVED** [review]

**Pull request merged**